### PR TITLE
[ITE-146] Update Segment Engage Destination to handle onDelete events

### PR DESCRIPTION
**Pull request by illumin04**
_A summary of your pull request, including the what change you're making and why._

Update Segment Engage Destination to handle onDelete events
Ticket link:[https://stackadapt.atlassian.net/browse/ITE-146](url)

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

No additional testing is done due to API not fully ready.
- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [ ] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/destination-actions/src/destinations/stackadapt-audiences/index.ts [hunk 1]
```diff
@@ -61,7 +133,45 @@ const destination: DestinationDefinition<Settings> = {
...
+          type: 'string',
+          required: false,
+          description: 'The anonymous ID to delete.'
+        },
+        advertiserId: {
+          label: 'Advertiser IDs',
+          type: 'string',
+          required: false,
+          description: 'Comma-separated list of advertiser IDs. If not provided, it will query the token info.'
+        },
```
Comment (Vanessa-SSY) [inline packages/destination-actions/src/destinations/stackadapt-audiences/index.ts:158]: is this suppose to be a list or a single advertiser id?

File: packages/destination-actions/src/destinations/stackadapt-audiences/index.ts [hunk 2]
```diff
@@ -61,7 +133,45 @@ const destination: DestinationDefinition<Settings> = {
   },
   actions: {
     forwardProfile,
-    forwardAudienceEvent
+    forwardAudienceEvent,
+    onDelete: {
```
Comment (Vanessa-SSY) [inline packages/destination-actions/src/destinations/stackadapt-audiences/index.ts:137]: I'd suggest refactor the onDelete to a separate folder like `forwardProfile` and `forwardAudienceEvent`

File: packages/destination-actions/src/destinations/braze/updateUserProfile/__tests__/__snapshots__/snapshot.test.ts.snap [hunk 3]
```diff
@@ -13,7 +13,7 @@ Object {
       },
       "date_of_first_session": "2021-02-01T00:00:00.000Z",
       "date_of_last_session": "2021-02-01T00:00:00.000Z",
-      "dob": "2021-02-01",
+      "dob": "2021-01-31",
```
Comment (Vanessa-SSY) [inline packages/destination-actions/src/destinations/braze/updateUserProfile/__tests__/__snapshots__/snapshot.test.ts.snap:16]: do we need this? Might need to rebase your local branch off of `main`: https://git-scm.com/docs/git-rebase

File: packages/destination-actions/src/destinations/optimizely-data-platform/__tests__/__snapshots__/snapshot.test.ts.snap [hunk 4]
```diff
@@ -112,8 +112,8 @@ Array [
     },
     "age": 89357760918978.56,
     "company": "WnJP3)h29qx6Q9XcA@qx",
-    "dob_day": 1,
-    "dob_month": 2,
+    "dob_day": 31,
+    "dob_month": 1,
```
Comment (Vanessa-SSY) [inline packages/destination-actions/src/destinations/optimizely-data-platform/__tests__/__snapshots__/snapshot.test.ts.snap:116]: same as above

File: packages/destination-actions/src/destinations/optimizely-data-platform/upsertContact/__tests__/__snapshots__/snapshot.test.ts.snap [hunk 5]
```diff
@@ -12,8 +12,8 @@ Array [
     },
     "age": 10705663206359.04,
     "company": "uMyg@6QjI31r!",
-    "dob_day": 1,
-    "dob_month": 2,
+    "dob_day": 31,
+    "dob_month": 1,
```
Comment (Vanessa-SSY) [inline packages/destination-actions/src/destinations/optimizely-data-platform/upsertContact/__tests__/__snapshots__/snapshot.test.ts.snap:16]: same as above. seems unrelated

Comment (illumin04) [reply packages/destination-actions/src/destinations/optimizely-data-platform/upsertContact/__tests__/__snapshots__/snapshot.test.ts.snap:16]: I will take at look at these unrelated changes, I remembered them changing when running all test or yarning something.

File: .husky/gitleaks-rules.toml [hunk 6]
```diff
@@ -12,6 +12,7 @@ paths = [
 regexes = ['''219-09-9999''', '''078-05-1120''', '''(9[0-9]{2}|666)-\d{2}-\d{4}''']
 
 [[rules]]
+id = "facebook_access_token"
```
[the remark this answers is not in the document]
Comment (illumin04) [reply .husky/gitleaks-rules.toml:15]: I added this because a test command showing in the repo can't run through, I don't think it impacts this section.

File: packages/destination-actions/src/destinations/stackadapt-audiences/__tests__/onDelete.test.ts [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    nock(gqlHostUrl)
+      .post(gqlPath)
+      .reply(200, {
+        data: {
+          tokenInfo: {
+            scopesByAdvertiser: {
+              nodes: []
+            }
+          }
+        }
```
Comment (Vanessa-SSY) [inline packages/destination-actions/src/destinations/stackadapt-audiences/__tests__/onDelete.test.ts:119]: I noticed that this shows up multiple times in the test. You can refactor it into a function and pass in the `nodes` part as input to reduce duplication

File: packages/destination-actions/src/destinations/stackadapt-audiences/__tests__/onDelete.test.ts [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+            userErrors {
+              message
+              path
+            }
+          }
+        }",
+      }
+    `)
+  })
+})
```
Comment (Vanessa-SSY) [inline packages/destination-actions/src/destinations/stackadapt-audiences/__tests__/onDelete.test.ts:259]: Great job! nice test cases!

File: packages/destination-actions/src/destinations/stackadapt-audiences/index.ts [hunk 9]
```diff
@@ -4,7 +4,79 @@ import type { Settings } from './generated-types'
 import forwardProfile from './forwardProfile'
 import forwardAudienceEvent from './forwardAudienceEvent'
 import { AdvertiserScopesResponse } from './types'
-import { GQL_ENDPOINT } from './functions'
+import { GQL_ENDPOINT, EXTERNAL_PROVIDER } from './functions'
```
Comment (Vanessa-SSY) [inline packages/destination-actions/src/destinations/stackadapt-audiences/index.ts:7]: Let's move this block for onDelete handling to its own folder. You can read about modularization: https://medium.com/@caitlinjeespn/modularization-in-software-engineering-1af52807ceed

File: packages/destination-actions/src/destinations/stackadapt-audiences/__tests__/onDelete.test.ts (hunk 7 above)
Comment (Vanessa-SSY) [reply packages/destination-actions/src/destinations/stackadapt-audiences/__tests__/onDelete.test.ts]: @illumin04 you might have missed this. Also applies to `deleteProfilesWithExternalIds ` query in test

File: packages/destination-actions/src/destinations/stackadapt-audiences/deleteProfile/__tests__/onDelete.test.ts [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+              externalProvider: \\"segmentio\\"
+            ) {
+              userErrors {
+                message
+                path
+              }
+            }
+          }",
+      }
+    `)
```
Comment (Vanessa-SSY) [inline packages/destination-actions/src/destinations/stackadapt-audiences/deleteProfile/__tests__/onDelete.test.ts:187]: can also refactor this to a helper

**Review by Vanessa-SSY — APPROVED** [review]
